hoomd/filter/filter_.py

```diff
@@ -11,4 +14,4 @@ def __str__(self):
 
     def __call__(self, state):
         """Needs to interact with state to get particles across MPI rank."""
-        raise NotImplementedError
+        return self._get_selected_tags(state._cpp_sys_def)
```

Comment:
this will fail on MPI, I think. The `ParticleFilters` were only designed to get the particles local to a rank. The union happens at the `ParticleGroup` level. We can change this of course. We could also just create a function in `hoomd.State` that grabs the particle group after updating as well.

Thoughts @joaander @vyasr?

Reply:
I don't think I know enough about how this part is planned to work going forward. It looks like there is now a single group class that is effectively constructed from a filter, is the idea that filters are more lightweight and easy to work with from python but groups must be constructed for practical usage with e.g. methods? I would avoid creating the kind of cyclic references it seems like would be introduced if you had to go Filter->State->Group->Filter, but I could be misunderstanding.

Reply:
The groups are required for C++. C++ groups own a filter. So the execution would go something like this Filter (Python) -> State (Python) -> Group (C++) -> Filter (C++) and then the C++ group would perform the union on all the rank specific tags from the filter.

The other option is to pass the MPI communicator with the state and make the call in Python `def __call__(self, simulation)`.

Reply:
We can either a) document that filters are performed locally only in conjunction with the to-be-implemented local state API. Or b) Implement a (possibly alternate) code path in ParticleFilter that performs the union from the rank outputs. The communicator needed to do this is already available in `state._simulation.device`. https://github.com/glotzerlab/hoomd-blue/blob/619f8a99dcead6f0118e09669b42999d6792d97e/hoomd/ParticleGroup.cc#L374-L397

Reply:
We could do something like this,

`def __call__(self, state, collect=False)`. Then we could default to local tags only, but we could provide an option to collect tags across ranks (we could limit this collecting to rank 0 if desired). If we do this, we should remove the logic in `ParticleGroup`s though.